### DISCUSSION TITLE
Thoughts on handling atomic memory orderings?

### DISCUSSION THREAD

**Discussion by alexrp** [phate/jlm#510] [Q&A]
Hi folks,

I'm curious if there's been any thought yet on how to model atomic instructions and memory orderings in the RVSDG? Concretely, for LLVM, we'd be talking about atomic `load`s/`store`s, `fence`, `cmpxchg`, and `atomicrmw`. ([LLVM has a handy doc on the supported memory orderings.](https://llvm.org/docs/Atomics.html))

Naïvely, it seems like it ought to be possible to represent these with memory state edges, but my thinking here is very surface-level at this stage. I'd love to hear any thoughts you have on this. 🙂

**Comment by phate** [comment]
I just had a quick look at these instructions, but have not thought very deeply about it. Memory edges is necessary for (some) of these instructions, but they are by themselves not sufficient. Some of these operations can be volatile. This means we effectively have an externally visible side-effect, which requires an I/O state. It might be that we can use specialized edges to encode the orderings of the atomic operations in the graph, but that would require some more thought and tinkering. There might be some research here, but I am far from an expert on this topic.

Ultimately, it would be possible to support them with just memory and I/O state edges. This would sequentialize them in the order that they were specified in the input LLVM file, but at the loss of optimization potential.

@haved @caleridas : Do you have an opinion on this matter?

Do you need support for them?

**Reply by alexrp** [reply]
> Do you need support for them?

I'm experimenting with applying the RVSDG + e-graphs as a C23 compiler IR. I'm not actually using JLM; I was just curious if you all had already thought about this problem since JLM is (probably?) the preeminent RVSDG research project. 🙂

> Ultimately, it would be possible to support them with just memory and I/O state edges. This would sequentialize them in the order that they were specified in the input LLVM file, but at the loss of optimization potential.

I think that's ok. I'm not too concerned with actually optimizing operations like this just yet, just making sure that they can actually be modeled correctly.

**Comment by caleridas** [comment]
For atomic data access itself, they can be modelled as mutations of the affected memory cells, so they would "look" very much the same as a store instruction.

A different story are the atomic fence operations: "acquire", "release", and "seq_cst". (Omitting "consume" memory ordering, as it is too weird, just assume it promoted to "acquire"): For e.g. acquire you need to ensure that every access to a (potentially shared) memory cell that happens in program order after the acquire operation is also sequenced after any preceding CPU atomic instruction. That in turn can also be modelled by making subsequent memory accesses dependent on the "acquire" operation by inserting suitable dependency edges. (Conversely for "release"). For seq_cst you would additional need global ordering of all seq_cst operations, so a single "seq_cst" state passed through everything (similar to the io state) accomplishes that.

So, in summary, I think there is almost nothing to do conceptually, just ensuring that proper dependency edges are inserted.

**Reply by alexrp** [reply]
> (Omitting "consume" memory ordering, as it is too weird, just assume it promoted to "acquire")

From what I've seen, committee members, compiler writers, and language users alike all seem to agree that `consume` was a mistake, so this seems reasonable. 😄

**Reply by caleridas** [reply]
Well I know why "consume" exists (I also wrote code for Dec Alpha, and that has something to do with it....). But defining a formal model for it quite a nightmare for the language front-end. Notably, you could actually totally sensibly do it in RVSDG as it is effectively after all a data flow representation (which is precisely what consume attempts to address), including precise tracking of the critical transformations that would need to be inhibited to prevent breaking the data dependency chain. That said, I think it is a lost cause.

**Reply by haved** [reply]
It seems LLVM only supports acquire, as that's what they use for C++'s `memory_order_consume`.

The LLVM atomics page has notes that "Optimizers not aware of atomics can treat this like a nothrow call." for all of acquire, release and seq_cst. A simple implementation for RVSDG based on how function calls are done, could thus be to merge all memory states, send it through the "function", which is the atomic operation, and then split them up again. This treats acquire and release as if they were fully seq_cst, but it should at least work.
